Improve `RedisCluster` slot caching and invalidation.

michael-grunder opened the pull request:
This PR contains a few changes to our slot caching mechanism and cluster remapping.

1. Elevate `CLUSTERDOWN` to an event that triggers a remapping of the keyspace. This can occur for example right after a node is removed from the cluster via `CLUSTER FORGET`. In addition, we now immediately remap the keyspace in this situation.
2. Add a debug function to allow the user to flush the persistent slot cache manually. Right now this would only affect the individual worker (or cli) process so may have limited utility initially.
3. Add a new `ini` setting to tell `RedisCluster` that the slot cache should expire after some number of seconds. This ensures that the keyspace is remapped even after non-destructive changes are made to the cluster such as a new replica being brought online. new replicas would never cause `MOVED` or `ASKING` responses so PhpRedis wouldn't know they were avaialble.

michael-grunder committed:
Rework how and when update slot cache and remap topology.

When a replica node is taken offline, it will briefly respond to
commands with some form of `CLUSTERDOWN` error. Presently while this did
cause PhpRedis to invalidate the slot cache, we did not immediately
remap the topology.

What this meant, is that for an already connected cluster, we might keep
sending readonly commands to this now decomissioned cluster replica for
the duration of the request.

Additionally, if the request was long lived enough we may still have it
mapped if the replica was brought back online at which point the client
might see dozens of `LOADING` exceptions.

This commit just changes our behavior when we detect any `CLUSTERDOWN`
error such that we flush the slot cache, disconnect the particular node,
and also attempt a remap of the topology.

michael-grunder committed:
Add a mechanism for the user to flush the slot cache.

michael-grunder committed:
Add a slot cache expiration mechanism

This commit adds a new `ini` setting `redis.clusters.slot_cache_expiry`
which specifies in seconds how long a cached slot map should live before
being refreshed. The setting is disabled by default to avoid breaking
backward compatibility.

michael-grunder committed:
Add `invalidateSlotCaches()` function to invalidate any slot caches.

This commit adds a mechanism to invalidate every `RedisCluster` slot
cache across all forked workers with one call to any of the workers.

The invalidation itslef works by using a `uint64_t` generation counter
which is allocated in shared memory on `MINIT` which is stored in every
slot cache when it is created.

When a user wants to invalidate any of the slot caches they can call
`RedisCluster::invalidateSlotCaches()` which simply does an atomic
increment on this global genertion value.

When a slot cache is retreived from the persistent list, the internal
generation is compared to the global shared memory generation and if
they don't match, the cache is not used, which is the same as
invalidting it as RedisCluster will write a new slot cache on object
destruction.

The feature is only compiled if `config.m4` determines that using `mmap`
to allocate shared memory is available and works.

Additionally, the feature must be explicitly enabled by setting the
`redis.clusters.shared_slot_cache_invalidation`.

yatsukhnenko commented on cluster_library.c:
```diff
@@ -883,6 +891,48 @@ cluster_free(redisCluster *c, int free_ctx)
+    if (expiry <= 0)
+        return 0;
+
+    return time(NULL) + expiry;
```
Shouldn't we use monotonic clock here?

michael-grunder replied:
Good point, probably. I'll need to see what the right way to do that is so it also compiles on Windows.

yatsukhnenko commented on cluster_library.c:
```diff
@@ -3105,22 +3157,31 @@ zend_string *cluster_hash_seeds(zend_string **seeds, uint32_t count) {
+    cc = le->ptr;
+    /* Short circuit if it should be expired */
+    if (cc->expiry != 0 && cc->expiry <= time(NULL))
+        return NULL;
```
I guess we need to remove element from `EG(persistent_list)` in this case and when generation mismatch

michael-grunder replied:
I think we get that for free with our dtor callback.

I can check though.

yatsukhnenko replied:
> I think we get that for free with our dtor callback.

Could you explain?

michael-grunder replied:
You're right. I thought that the destruction would happen automatically when the cache is replaced like other zend hash tables do.

michael-grunder committed:
Use zend_hrtime instead of time(NULL) since it's monotonic

`time(NULL)` isn't monotonic and can run backwards so switch to
`zend_hrtime` which uses `clock_gettime(CLOCK_MONOTONIC)` on Linux and
`QueryPerformanceCounter`on Windows.

michael-grunder committed:
Clean up invalidated slot cache
